pkg/apis/resource/validation/validation.go

```diff
@@ -703,35 +703,23 @@ func validateResourceSliceSpec(spec, oldSpec *resource.ResourceSliceSpec, fldPat
 			"exactly one of `nodeName`, `nodeSelector`, `allNodes`, `perDeviceNodeSelection` is required, but multiple fields are set"))
 	}
 
-	sharedCounterToCounterNames := gatherSharedCounterCounterNames(spec.SharedCounters)
-	allErrs = append(allErrs, validateSet(spec.Devices, resource.ResourceSliceMaxDevices,
+	if spec.SharedCounters != nil && spec.Devices != nil {
+		allErrs = append(allErrs, field.Invalid(fldPath, "", "only one of `sharedCounters` or `devices` is allowed"))
+	}
+
+	maxDevices := resource.ResourceSliceMaxDevices
+	if haveConsumesCounters(spec) {
+		maxDevices = resource.ResourceSliceMaxDevicesWithConsumesCounters
```

Comment:
same comment as at https://github.com/kubernetes/kubernetes/pull/134152#discussion_r2486630747

once this PR and that PR need to both live in master, should we combine their device count limit detection into a single function with a generic constant / name like `ResourceSliceMaxDevicesWithComplexFields` or something?

Reply:
+1

Reply:
I agree that we can combine both features with a limit of 64, assuming that total device size doesn't get too big - to be verified once the first feature is merged and while adapting the second.

Reply:
Yeah, I've updated this to follow the same pattern as in https://github.com/kubernetes/kubernetes/pull/134152 for now. But it needs to be updated if https://github.com/kubernetes/kubernetes/pull/134152 merges first (which I think it probably will).

I don't think the size will be a problem, but we definitely need to verify.